// pepesan/dist/Types/Response/Audio.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BufferAudio = exports.URLAudio = exports.Audio = void 0;
const _1 = require(".");
class Audio extends _1.MessageResponse {
    constructor(ptt) {
        super();
        this.ptt = ptt !== null && ptt !== void 0 ? ptt : false;
    }
    static fromURL(url, ptt) {
        return new URLAudio(url, ptt);
    }
    static fromBuffer(buffer, ptt) {
        return new BufferAudio(buffer, ptt);
    }
    getMessageContent() {
        return;
    }
}
exports.Audio = Audio;
class URLAudio extends Audio {
    constructor(url, ptt) {
        super(ptt);
        this.url = url;
    }
    getMessageContent() {
        return { audio: { url: this.url }, mimetype: 'audio/mp4' };
    }
}
exports.URLAudio = URLAudio;
class BufferAudio extends Audio {
    constructor(buffer, ptt) {
        super(ptt);
        this.buffer = buffer;
    }
    getMessageContent() {
        return { audio: this.buffer, mimetype: 'audio/mp4', ptt: this.ptt };
    }
}
exports.BufferAudio = BufferAudio;
