// pepesan/dist/Types/Response/Document.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BufferDocument = exports.URLDocument = exports.Document = void 0;
const _1 = require(".");
class Document extends _1.MessageResponse {
    constructor(mimetype, fileName) {
        super();
        this.mimetype = mimetype !== null && mimetype !== void 0 ? mimetype : '';
        this.fileName = fileName !== null && fileName !== void 0 ? fileName : 'document';
    }
    static fromURL(url, mimetype, fileName) {
        return new URLDocument(url, mimetype, fileName);
    }
    static fromBuffer(buffer, mimetype, fileName) {
        return new BufferDocument(buffer, mimetype, fileName);
    }
    getMessageContent() {
        return;
    }
}
exports.Document = Document;
class URLDocument extends Document {
    constructor(url, mimetype, fileName) {
        super(mimetype, fileName);
        this.url = url;
    }
    getMessageContent() {
        return { document: { url: this.url }, mimetype: this.mimetype };
    }
}
exports.URLDocument = URLDocument;
class BufferDocument extends Document {
    constructor(buffer, mimetype, fileName) {
        super(mimetype, fileName);
        this.buffer = buffer;
    }
    getMessageContent() {
        return { document: this.buffer, mimetype: this.mimetype, fileName: this.fileName };
    }
}
exports.BufferDocument = BufferDocument;
