// pepesan/dist/Types/Response/Image.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BufferImage = exports.URLImage = exports.Image = void 0;
const _1 = require(".");
class Image extends _1.MessageResponse {
    constructor(caption) {
        super();
        this.caption = caption;
    }
    static fromURL(url, caption) {
        return new URLImage(url, caption);
    }
    static fromBuffer(buffer, caption) {
        return new BufferImage(buffer, caption);
    }
    getMessageContent() {
        return;
    }
}
exports.Image = Image;
class URLImage extends Image {
    constructor(url, caption) {
        super(caption);
        this.url = url;
    }
    getMessageContent() {
        return { image: { url: this.url }, caption: this.caption };
    }
}
exports.URLImage = URLImage;
class BufferImage extends Image {
    constructor(buffer, caption) {
        super(caption);
        this.buffer = buffer;
    }
    getMessageContent() {
        return { image: this.buffer, caption: this.caption };
    }
}
exports.BufferImage = BufferImage;
