// pepesan/dist/Types/Chat.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Response = void 0;
const Response_1 = require("./Response");
class Response {
    static getResponseClass(responseClass) {
        responseClass.prototype.clientId = this.clientId;
        return responseClass;
    }
    static get image() {
        return this.getResponseClass(Response_1.Image);
    }
    static get text() {
        return this.getResponseClass(Response_1.Text);
    }
    static get sticker() {
        return this.getResponseClass(Response_1.Sticker);
    }
    static get audio() {
        return this.getResponseClass(Response_1.Audio);
    }
    static get video() {
        return this.getResponseClass(Response_1.Video);
    }
    static get button() {
        return this.getResponseClass(Response_1.Button);
    }
    static get list() {
        return this.getResponseClass(Response_1.List);
    }
    static get menu() {
        return this.getResponseClass(Response_1.MenuResponse.Menu);
    }
    static get document() {
        return this.getResponseClass(Response_1.Document);
    }
}
exports.Response = Response;
