// pepesan/dist/Database/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Database = void 0;
const sequelize_1 = require("sequelize");
const PepesanModels = __importStar(require("./Models"));
class Database {
    constructor(dbConfig, models) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        this.name = (_a = dbConfig === null || dbConfig === void 0 ? void 0 : dbConfig.name) !== null && _a !== void 0 ? _a : 'pepesan';
        this.user = (_b = dbConfig === null || dbConfig === void 0 ? void 0 : dbConfig.user) !== null && _b !== void 0 ? _b : 'admin';
        this.pass = (_c = dbConfig === null || dbConfig === void 0 ? void 0 : dbConfig.pass) !== null && _c !== void 0 ? _c : '#p3p3s4n!';
        this.path = (_d = dbConfig === null || dbConfig === void 0 ? void 0 : dbConfig.path) !== null && _d !== void 0 ? _d : './data.sqlite';
        this.dialect = (_e = dbConfig === null || dbConfig === void 0 ? void 0 : dbConfig.dialect) !== null && _e !== void 0 ? _e : 'sqlite';
        this.host = (_f = dbConfig === null || dbConfig === void 0 ? void 0 : dbConfig.host) !== null && _f !== void 0 ? _f : 'localhost';
        this.port = (_g = dbConfig === null || dbConfig === void 0 ? void 0 : dbConfig.port) !== null && _g !== void 0 ? _g : 3306;
        this.syncAlter = (_h = dbConfig === null || dbConfig === void 0 ? void 0 : dbConfig.syncAlter) !== null && _h !== void 0 ? _h : false;
        this.timezone = dbConfig === null || dbConfig === void 0 ? void 0 : dbConfig.timezone;
        this.models = models !== null && models !== void 0 ? models : [];
        this.initSql();
        this.initModels();
    }
    initSql() {
        this.sql = new sequelize_1.Sequelize({
            dialect: this.dialect,
            host: this.host,
            port: this.port,
            database: this.name,
            username: this.user,
            password: this.pass,
            storage: this.path,
            timezone: this.timezone,
            define: {
                underscored: true
            }
        });
    }
    initModels() {
        const sqlModels = [...Object.values(PepesanModels), ...this.models];
        const sql = this.sql;
        for (const model of sqlModels) {
            // @ts-ignore
            model.init(model.attributes, {
                sequelize: sql
            });
            model.sync({ alter: this.syncAlter });
        }
    }
}
exports.Database = Database;
__exportStar(require("./Models"), exports);
